### PR TITLE
Ruler external support

### LINKED ISSUE #4782: Cross cluster rule evaluation
***Is your feature request related to a problem? Please describe.***
Currently, Cortex rulers are limited to their own Cortex cluster. They can only evaluate rules and alert from the Ingesters of their own cell. This causes limited flexibility in how Cortex clusters can be run.

***Describe the solution you'd like***
In order to allow Cortex Rulers to communicate with other cortex clusters, we should allow custom implementations of the Appender and Querier. These can be passed in to Cortex through a newly introduced config. Using these custom implementations, the Ruler can remote-write and remote-read to other external endpoints. This would allow for cortex clusters to used solely for rule evaluation and alerting and vice versa others for just ingestion and querying. (see picture). 
Github Issue (1)

These custom implementations of Appender and Querier are not necessary and default implementations will still be able to be used if necessary.

### PR DESCRIPTION
**Pull request by jmoscoso1**
**What this PR does**:
Adds support for custom implementations of pusher and queryable

Example of custom implementation of pusher and queryable being used:
```
LoadConfig(configFile, expandENV, &cfg)

cfg.ExternalQueryable = querier.NewQueryable(c)
cfg.ExternalPusher = pusher.NewPusher(w)

t, err := cortex.New(cfg)

err = t.Run()
```
Full code at https://github.com/jmoscoso1/cortex_external_ruler
and the docker compose environment where it is run https://github.com/jmoscoso1/Docker_compose_cortex

**Which issue(s) this PR fixes**:
Fixes #4782 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Changelog
 
 ## master / unreleased
-
+* [FEATURE] Ruler: Add support for external implementations of Queryable and Pusher #4782
```
Comment (harry671003) [inline CHANGELOG.md:4]: nit: Add support to pass custom implementations of queryable and pusher.

File: pkg/cortex/cortex.go [hunk 2]
```diff
@@ -92,6 +92,9 @@ type Config struct {
 	PrintConfig bool                   `yaml:"-"`
 	HTTPPrefix  string                 `yaml:"http_prefix"`
 
+	ExternalQueryable		prom_storage.Queryable		`yaml:"-"`
```
Comment (harry671003) [inline pkg/cortex/cortex.go:95]: Are these files correctly formatted?
`gofmt -f -s -w <PATH>`

File: pkg/cortex/modules.go [hunk 3]
```diff
@@ -74,6 +75,7 @@ const (
 	TableManager             string = "table-manager"
 	RulerStorage             string = "ruler-storage"
 	Ruler                    string = "ruler"
+	RulerExternal			 string = "ruler-external"
```
Comment (harry671003) [inline pkg/cortex/modules.go:78]: nit: Formatting looks off.

File: pkg/cortex/modules.go [hunk 4]
```diff
@@ -652,6 +654,66 @@ func (t *Cortex) initRulerStorage() (serv services.Service, err error) {
...
+	return nil
+}
+
+func (t *Cortex) initRulerExternal() (serv services.Service, err error) {
+	if t.RulerStorage == nil {
+		level.Info(util_log.Logger).Log("msg", "RulerStorage is nil.  Not starting the ruler.")
+		return nil, nil
+	}
+	cfg := t.Cfg.Querier
+	logger := util_log.Logger
```
Comment (harry671003) [inline pkg/cortex/modules.go:674]: nit: These assignments seems unnecessary.

File: pkg/cortex/modules.go [hunk 5]
```diff
@@ -652,6 +654,66 @@ func (t *Cortex) initRulerStorage() (serv services.Service, err error) {
...
+		t.Cfg.Ruler,
+		manager,
+		prometheus.DefaultRegisterer,
+		util_log.Logger,
+		t.RulerStorage,
+		t.Overrides,
+	)
+
+	if err != nil {
+		return
```
Comment (harry671003) [inline pkg/cortex/modules.go:703]: Should be: `return nil, error`

File: pkg/cortex/cortex.go [hunk 6]
```diff
@@ -92,6 +92,9 @@ type Config struct {
...
+	ExternalPusher 			ruler.Pusher				`yaml:"-"`
```
Comment (harry671003) [inline pkg/cortex/cortex.go:96]: Could you also add links to where this is used in your POC project?

Also add more details in the PR description about how we can initialize a Cortex instance with External implementation.

File: pkg/cortex/modules.go [hunk 7]
```diff
@@ -652,6 +654,64 @@ func (t *Cortex) initRulerStorage() (serv services.Service, err error) {
...
+		MaxSamples:         t.Cfg.Querier.MaxSamples,
+		Timeout:            t.Cfg.Querier.Timeout,
+		LookbackDelta:      t.Cfg.Querier.LookbackDelta,
+		EnablePerStepStats: t.Cfg.Querier.EnablePerStepStats,
+		EnableAtModifier:   t.Cfg.Querier.AtModifierEnabled,
+		NoStepSubqueryIntervalFn: func(int64) int64 {
+			return t.Cfg.Querier.DefaultEvaluationInterval.Milliseconds()
+		},
+	})
+	managerFactory := ruler.DefaultTenantManagerFactory(t.Cfg.Ruler, t.Cfg.ExternalPusher, t.Cfg.ExternalQueryable, engine, t.Overrides, prometheus.DefaultRegisterer)
```
Comment (harry671003) [inline pkg/cortex/modules.go:686]: Before we pass the ExternalPusher and ExternalQueryable, I think we should validate that those are not nulls.

File: pkg/cortex/modules.go [hunk 8]
```diff
@@ -907,6 +943,9 @@ func (t *Cortex) setupModuleManager() error {
 		TenantFederation:         {Queryable},
 		All:                      {QueryFrontend, Querier, Ingester, Distributor, TableManager, Purger, StoreGateway, Ruler},
 	}
+	if t.Cfg.ExternalPusher != nil && t.Cfg.ExternalQueryable != nil {
```
Comment (harry671003) [inline pkg/cortex/modules.go:946]: If possible add a unit test for this.

File: pkg/cortex/modules.go [hunk 9]
```diff
@@ -652,19 +653,54 @@ func (t *Cortex) initRulerStorage() (serv services.Service, err error) {
...
+				return t.Cfg.Querier.DefaultEvaluationInterval.Milliseconds()
+			},
+		})
+
+		managerFactory := ruler.DefaultTenantManagerFactory(t.Cfg.Ruler, t.Cfg.ExternalPusher, t.Cfg.ExternalQueryable, engine, t.Overrides, prometheus.DefaultRegisterer)
+		manager, err = ruler.NewDefaultMultiTenantManager(t.Cfg.Ruler, managerFactory, prometheus.DefaultRegisterer, util_log.Logger)
+
+	} else {
+
+		t.Cfg.Ruler.Ring.ListenPort = t.Cfg.Server.GRPCListenPort
```
Comment (alanprot) [inline pkg/cortex/modules.go:694]: Dont we need to do this on both cases?

File: pkg/cortex/modules.go (hunk 8 above)
Comment (jmoscoso1) [reply pkg/cortex/modules.go:946]: that checks dependencies?

**Review by alanprot — APPROVED** [review]

**Comment by alanprot** [comment]
Build and Lint not working! =/

**Review by alvinlin123 — APPROVED** [review]

**Pull request merged**